Editorial review: Document the ToggleEvent.source property (#40533)
mdn/content: files/en-us/web/api/toggleevent/source/index.md

---
title: "ToggleEvent: source property"
short-title: source
slug: Web/API/ToggleEvent/source
page-type: web-api-instance-property
browser-compat: api.ToggleEvent.source
---

{{APIRef("Popover API")}}

The **`source`** read-only property of the {{domxref("ToggleEvent")}} interface is an {{domxref("Element")}} object instance representing the HTML popover control element that initiated the toggle.

> A {{htmlelement("button")}} or `<input type="button">` element can be specified as a popover control by specifying the [`id`](/en-US/docs/Web/HTML/Reference/Global_attributes/id) of the popover element in its [`commandfor`](/en-US/docs/Web/HTML/Reference/Elements/button#commandfor) or [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget) attribute.

Author: Yeah, I think you are right. I've clarified this like so:

> A {{htmlelement("button")}} element can be specified as a popover control by specifying the [`id`](/en-US/docs/Web/HTML/Reference/Global_attributes/id) of the popover element in its [`commandfor`](/en-US/docs/Web/HTML/Reference/Elements/button#commandfor) or [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget) attribute (if the button is specified using `<input type="button">`, only the `popovertarget` attribute works).

> A {{htmlelement("button")}} element can be specified as a popover control by specifying the [`id`](/en-US/docs/Web/HTML/Reference/Global_attributes/id) of the popover element in its [`commandfor`](/en-US/docs/Web/HTML/Reference/Elements/button#commandfor) or [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget) attribute (if the button is specified using `<input type="button">`, only the `popovertarget` attribute works).

Reviewer: This doesn't feel "properly associated" with the preceding statement. 
```suggestion
The {{htmlelement("button")}} element that is in control is the [`id`](/en-US/docs/Web/HTML/Reference/Global_attributes/id) of the popover element in its [`commandfor`](/en-US/docs/Web/HTML/Reference/Elements/button#commandfor) or [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget) attribute (if the button is specified using `<input type="button">`, only the `popovertarget` attribute works).
```

  <button id="no" commandfor="popover" command="hide-popover">No</button>
</div>
```

```css hidden live-sample___toggleevent-source
html {
  font-family: sans-serif;
}

[popover] {
  border: 1px solid grey;
  padding: 10px 20px;
  border-radius: 5px;
}

[popover] h3 {
  margin: 0 0 10px;
}
```

#### JavaScript

In our script, we start off by grabbing references to the "yes" and "no" buttons, the popover, and the output `<p>`.

```js live-sample___toggleevent-source
const yesBtn = document.getElementById("yes");
const noBtn = document.getElementById("no");
const popover = document.getElementById("popover");
const output = document.getElementById("output");
```

We then add a [`toggle`](/en-US/docs/Web/API/HTMLElement/toggle_event) event listener to the popover. When fired, it checks if the "yes" or the "no" button was used to toggle (hide) the popover; an appropriate message is printed to the output `<p>` in each case.

```js live-sample___toggleevent-source
popover.addEventListener("toggle", (event) => {
  if (event.source === yesBtn) {
    output.textContent = "Cookie set!";
  } else if (event.source === noBtn) {
    output.textContent = "No cookie set.";

>   }

Reviewer: This doesn't work on any current browser (I just updated and 138 is the latest being fetched). 
Lets add a feature check

```suggestion
  } else {
      output.textContent = "ToggleEvent.source not supported.";
  }
```

Author: Good call. I've added feature detection and updated the surrounding description. Note that this ended up being a bit more complicated than first expected, because we need to check for `command` support too.
